presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -231,4 +233,9 @@ public TaskContext addTaskContext(TaskStateMachine taskStateMachine, Session ses
                 .map(taskContext -> taskContext.accept(visitor, context))
                 .collect(toList());
     }
+
+    public Optional<TaskContext> getTaskContextByTaskId(TaskId taskId)
+    {
+        return taskContexts.stream().filter(context -> context.getTaskId().equals(taskId)).findFirst();
```

Comment:
This inevitably is linear. What about key-ing tasks contexts by their ID (maybe as a separate structure in this class)? Or persisting returned value in `UpdateSystemMemory`?